src/main/kotlin/org/rust/ide/annotator/CargoCheckExtAnnotator.kt

```diff
@@ -0,0 +1,630 @@
+                                    "warning" -> return WarningMessage(children, code, message, spans)
+                                    else -> throw AssertionError()
+                                }
+
+                            }
+                            "help" -> return HelpMessage(message, spans)
+                            "note" -> return NoteMessage(message, spans)
+
+                            else -> throw RuntimeException("Unrecognized level `$level` in rust compiler message")
+                        }
+                    }
+                }
+
+            }
+
+            //An error code (?)
+            data class Code(val code : String, val explanation: String?) {
+
+                fun formatAsLink() : String {
+                    val rustErrorIndexPrefix = "https://doc.rust-lang.org/error-index.html"
+
+                    return "<a href=\"$rustErrorIndexPrefix#$code\">$code</a>"
+                }
+            }
+
+            //Code spans (?)
+            data class Span(
+                val byte_end : Int,
+                val byte_start : Int,
+                val column_end : Int,
+                val column_start : Int,
+                val expansion : Expansion?,
+                val file_name : String,
+                val is_primary : Boolean,
+                val label : String?,
+                val line_end : Int,
+                val line_start : Int,
+                //May be JsonNull
+                val suggested_replacement : JsonElement,
+                val text : List<Text>
+            ) {
+
+                //This can be used to get all children spans (e.g. in expansion) including
+                //this one.
+                fun getAllSpans() : List<Span> {
+                    return listOf(this) +
+                        (this.expansion?.def_site_span?.getAllSpans() ?: listOf()) +
+                        (this.expansion?.span?.getAllSpans() ?: listOf())
+
+                }
+
+                //A mysterious field that is used with macros it seems
+                //Unfortunately it may hold valid spans (and sometimes it
+                //holds the only valid span of a message) so we have to
+                //parse them.
+                data class Expansion(
+                    //May be null
+                    val def_site_span : Span?,
+                    val macro_decl_name : String,
+                    val span : Span
+                )
+
+                data class Text(
+                    val highlight_end : Int,
+                    val highlight_start : Int,
+                    val text : String
+                )
+
+            }
+
+        }
+
+        //file - The file to annotate
+        fun doAnnotate(holder: AnnotationHolder)  {
+            message.doAnnotate(holder, target)
+        }
+    }
+
+
+}
+
+data class TEAAnnotationInfo(val file: PsiFile, val editor: Editor)
```

Comment:
It's better to avoid acronyms in names, unless the acronym is very popular, like HTML.